### PR TITLE
feat(storage): add redis storage

### PR DESCRIPTION
**Pull request by nacknime-official**
This PR is based on #2!

The storage based on JSON like aiogram.

TODO: think about `context.TODO()`, maybe we can add a new argument `ctx` into each storage method?
But anyway `telebot` doesn't support context.Context "flow" :-(

### CODE REVIEW

**Comment by nacknime-official** [comment]
Oh, I found out that for the numbers `json.Unmarshal` always returns `float64`.
Even if we use `decoder.UseNumbers()`, we can return only int64 or float64, not int32, int16 etc, so we can't return the initial type for the integers or floats.

The solution might be storing the type of the variable as well as the value just like:
```json
{
  "inputName": {"type": "string", "value": "Peter"},
  "inputAge":  {"type": "int", "value": 21}
}
```

But then we don't even need to store the state data in JSON I think.

**Comment by nacknime-official** [comment]
> The solution might be storing the type of the variable as well as the value just like:


Seems that will be so hard and even impossible for some custom types I think.

Seems the only one easy way to solve it - encode the value via "encoding/gob" and let the "GetData" method accept "to interface{}" argument and not return the value like

```go
GetData(chatId, userId int64, key string, to interface{}) error
```

and the client code will look like
```go
var age int
if err := state.Get("age", &age); err != nil {
  return err
}
fmt.Println(age) // 21
```

It needs to change the `Context` and `Storage` interface, but I don't see any other type-safe solution.

**Comment by nacknime-official** [comment]
Here is the example of letting the "GetData" save the data to "to interface{}" argument: https://github.com/vitaliy-ukiru/fsm-telebot/pull/4

File: storages/redis/redis.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	// Default is 0.
+	ttlState time.Duration
+
+	// TTL for state data.
+	// Default is 0.
+	ttlData time.Duration
+}
+
+// NewStorage returns new redis storage.
+func NewStorage(client *redis.Client, pref StorageSettings) fsm.Storage {
```
Comment (vitaliy-ukiru) [inline storages/redis/redis.go:44]: make return type "*Storage" (i' ll change it in memory storage also)

File: storages/redis/redis.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	prefix string
+
+	// TTL for state.
+	// Default is 0.
+	ttlState time.Duration
+
+	// TTL for state data.
+	// Default is 0.
+	ttlData time.Duration
+}
```
Comment (vitaliy-ukiru) [inline storages/redis/redis.go:41]: the fields are non-exportable, you cannot set the value by key

File: go.mod [hunk 3]
```diff
@@ -4,7 +4,13 @@ go 1.17
 
 require (
 	github.com/pkg/errors v0.9.1
+	github.com/redis/go-redis/v9 v9.0.2
```
Comment (vitaliy-ukiru) [inline go.mod:7]: create new module in storages/redis, not everyone will use this dependency to place it here

File: storages/redis/redis.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	return nil
+}
+
+func (s *Storage) UpdateData(chatId, userId int64, key string, data interface{}) error {
+	stateDataBytes, err := s.rds.Get(context.TODO(), s.generateKey(chatId, userId, stateDataKey)).Bytes()
+	if err != nil {
+		if errors.Is(err, redis.Nil) {
+			stateDataBytes = []byte("{}")
+		} else {
+			return fmt.Errorf("get state data: %w", err)
```
Comment (vitaliy-ukiru) [inline storages/redis/redis.go:90]: add prefix like "fsm-telebot/storage/redis: UpdateData: ". May create function for this

Comment (vitaliy-ukiru) [inline storages/redis/redis.go:90]: I also use "github.com/pkg/errors" to wrap the errors. The code style is different here. Maybe I'll change that later.

**Review by vitaliy-ukiru — CHANGES_REQUESTED** [review]

File: storages/redis/redis.go (hunk 2 above)
Comment (nacknime-official) [reply storages/redis/redis.go:41]: really, my bad)

**Comment by vitaliy-ukiru** [comment]
> TODO: think about context.TODO(), maybe we can add a new argument ctx into each storage method?

good idea, I think can add the `WithContext(context.Context) fsm.Context` method to `fsm.Context`, default sets `context.Background`, or (and) embed `context.Context`

File: storages/redis/redis.go (hunk 1 above)
Comment (nacknime-official) [reply storages/redis/redis.go:44]: resolved in #4 https://github.com/vitaliy-ukiru/fsm-telebot/pull/4/commits/5a5f86b62b6da9b1bbe14eb23a4b54649bef6e93

File: storages/redis/redis.go (hunk 2 above)
Comment (nacknime-official) [reply storages/redis/redis.go:41]: resolved in #4 https://github.com/vitaliy-ukiru/fsm-telebot/pull/4/commits/5a5f86b62b6da9b1bbe14eb23a4b54649bef6e93

File: go.mod (hunk 3 above)
Comment (nacknime-official) [reply go.mod:7]: resolved in #4 https://github.com/vitaliy-ukiru/fsm-telebot/pull/4/commits/90053aaa09a7426ec1ead39c4ee0fc7e433b8ae7

File: storages/redis/redis.go (hunk 4 above)
Comment (nacknime-official) [reply storages/redis/redis.go:90]: resolved in #4 https://github.com/vitaliy-ukiru/fsm-telebot/pull/4/commits/ff6dd0bba0947b21b7ad6cb27363284c3e59440c

Comment (nacknime-official) [reply storages/redis/redis.go:90]: resolved in #4 https://github.com/vitaliy-ukiru/fsm-telebot/pull/4/commits/ff6dd0bba0947b21b7ad6cb27363284c3e59440c

**Comment by nacknime-official** [comment]
> > TODO: think about context.TODO(), maybe we can add a new argument ctx into each storage method?
> 
> good idea, I think can add the `WithContext(context.Context) fsm.Context` method to `fsm.Context`, default sets `context.Background`, or (and) embed `context.Context`

if you mean to put the client's context into the `fsm.Context` as a field of the `fsmContext` structure, then don't know, what if the client wants to control every `Set` and `Get` with his own contexts

**Comment by nacknime-official** [comment]
> I don't want "FSMContext" to always require context, this will make interface more complex. I want it to be light version of access to storage.

but it's the usual way in Go, isn't it? :)

**Pull request merged**